src/mpid/ch4/src/ch4r_symheap.h

```diff
@@ -193,6 +193,309 @@ static inline void *MPIDI_CH4R_generate_random_addr(size_t size)
+
+        /* unlink shared memory region so it gets deleted when all processes exit */
+        mpl_err = MPL_shm_seg_remove(*shm_segment_hdl_ptr);
+        MPIR_ERR_CHKANDJUMP(mpl_err != MPL_SHM_SUCCESS, mpi_errno, MPI_ERR_OTHER,
+                            "**remove_shar_mem");
+    } else {
+        /* receive the mapping result of rank 0 */
+        mpi_errno = MPIR_Bcast(mapfail_flag, 1, MPI_INT, 0, shm_comm_ptr, &errflag);
+
+        /* if rank 0 mapped successfully, others on the node attach shared memory region */
+        if (!*mapfail_flag) {
+            char serialized_hnd[MPL_SHM_GHND_SZ] = { 0 };
+
+            /* get serialized handle from rank 0 and deserialize it */
+            mpi_errno = MPIR_Bcast(serialized_hnd, MPL_SHM_GHND_SZ, MPI_CHAR, 0,
+                                   shm_comm_ptr, &errflag);
+            MPIR_ERR_CHKANDJUMP(errflag, mpi_errno, MPI_ERR_OTHER, "**coll_fail");
+
+            mpl_err = MPL_shm_hnd_deserialize(*shm_segment_hdl_ptr, serialized_hnd,
+                                              strlen(serialized_hnd));
+            MPIR_ERR_CHKANDJUMP(mpl_err != MPL_SHM_SUCCESS, mpi_errno, MPI_ERR_OTHER,
+                                "**alloc_shar_mem");
+
+            /* try to attach with specified symmetric address */
+            mpl_err = MPL_shm_fixed_seg_attach(*shm_segment_hdl_ptr, shm_segment_len, base_ptr, 0);
+            if (mpl_err == MPL_SHM_EINVAL)
+                *mapfail_flag = 1;
+
+            /* handle mapping failure separately, caller may retry predefined times before giving up */
+            MPIR_ERR_CHKANDJUMP((mpl_err != MPL_SHM_EINVAL && mpl_err != MPL_SHM_SUCCESS),
+                                mpi_errno, MPI_ERR_OTHER, "**alloc_shar_mem");
+
+            mpi_errno = MPIR_Barrier(shm_comm_ptr, &errflag);
+            MPIR_ERR_CHKANDJUMP(errflag, mpi_errno, MPI_ERR_OTHER, "**coll_fail");
+        }
+    }
+
+  fn_exit:
+    MPIR_FUNC_VERBOSE_EXIT(MPID_STATE_MPIDI_CH4I_ALLOCATE_SYMSHM_SEGMENT);
+    return mpi_errno;
+  fn_fail:
+    goto fn_exit;
+}
+
+typedef struct {
+    unsigned long long sz;
+    int loc;
+} MPIDI_CH4I_ull_maxloc_t;
+
+/* Compute maxloc for unsigned long long type.
+ * If more than one max value exists, the loc with lower rank is returned. */
+static inline void MPIDI_CH4I_ull_maxloc_op_func(void *invec, void *inoutvec,
+                                                 int *len, MPI_Datatype * datatype)
+{
+    MPIDI_CH4I_ull_maxloc_t *inmaxloc = (MPIDI_CH4I_ull_maxloc_t *) invec;
+    MPIDI_CH4I_ull_maxloc_t *outmaxloc = (MPIDI_CH4I_ull_maxloc_t *) inoutvec;
+    int i;
+    for (i = 0; i < *len; i++) {
+        if (inmaxloc->sz > outmaxloc->sz) {
+            outmaxloc->sz = inmaxloc->sz;
+            outmaxloc->loc = inmaxloc->loc;
+        } else if (inmaxloc->sz == outmaxloc->sz && inmaxloc->loc < outmaxloc->loc)
+            outmaxloc->loc = inmaxloc->loc;
+    }
+}
+
+/* Allredue MAXLOX for unsigned size type by using user defined operator
```

Comment:
Allreduce MAXLOC

Also let's mention why we needed this i.e. standard `MAXLOC` doesn't support unsigned types, thus not compatible with `size_t`.

Reply:
Sounds good to me. I just pushed the improved comment. Hopefully it does not interrupt the tests.